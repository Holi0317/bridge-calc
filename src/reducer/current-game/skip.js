// @flow
import mapValues from 'lodash-es/mapValues'
import {GameStage} from '../../game-stage'
import {toFront} from '../../utils'
import {toEndedState} from './converter'
import {bidWinGenerator} from './bid-win-generator'

import type {EndedState, WaitingBidState, WaitingWinState} from './types'

/**
 * Helper function for reducer.
 * Skip n round(s) of game.
 */
export function skip(state: WaitingBidState | WaitingWinState, n: number, time: Date): WaitingBidState | EndedState | null {
  if (state === null) {
    return null
  }
  if (n === 0) {
    // Last skip action.
    if (state.currentRound === state.rounds) {
      // Last round
      return toEndedState(state, time)
    } else {
      // Not last round
      return state
    }
  }
  const newState: WaitingBidState = {
    ...state,
    stage: GameStage.waitingBid,
    bid: bidWinGenerator(Object.keys(state.names)),
    scores: mapValues(state.scores, score => [...score, 0]),
    currentPlayerOrder: toFront(state.currentPlayerOrder, 1),
    currentRound: state.currentRound + 1
  }

  // Remove unused property
  delete (newState: any).win

  return skip(newState, n - 1, time)
}
