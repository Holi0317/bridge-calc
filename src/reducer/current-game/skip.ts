import mapValues from 'lodash-es/mapValues'
import {toFront} from '../../utils'
import {toEndedState, toWaitingBidState} from './converter'
import {bidWinGenerator} from './bid-win-generator'
import {IEndedState, IWaitingBidState, IWaitingWinState} from './types'

/**
 * Helper function for reducer.
 * Skip n round(s) of game.
 */
export function skip(state: IWaitingBidState | IWaitingWinState, n: number, time: Date): IWaitingBidState | IEndedState | null {
  if (state === null) {
    return null
  }
  if (n === 0) {
    // Last skip action.
    if (state.currentRound === state.rounds) {
      // Last round
      return toEndedState(state, time)
    } else {
      // Not last round
      return toWaitingBidState(state)
    }
  }

  const newState = toWaitingBidState(state)
  newState.bid = bidWinGenerator(Object.keys(state.names))
  newState.scores = mapValues(state.scores, score => [...score, 0])
  newState.currentPlayerOrder = toFront(state.currentPlayerOrder, 1)
  newState.currentRound = state.currentRound + 1

  return skip(newState, n - 1, time)
}
