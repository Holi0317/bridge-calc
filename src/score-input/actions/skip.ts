import {ActionTypes} from '../../action-types'

export interface ISkipAction {
  type: ActionTypes.SKIP,
  times?: number,
  time: Date
}

/**
 * Skip n round(s).
 * @param times - Number of rounds to be skip.
 * If not provided, current round will be skipped.
 */
export function skipAction(times?: number): ISkipAction {
  return {type: ActionTypes.SKIP, times, time: new Date()}
}
