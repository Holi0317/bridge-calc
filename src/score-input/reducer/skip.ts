import mapValues from 'lodash-es/mapValues'
import {toFront} from '../../utils'
import {toEndedState, toWaitingBidState} from './converter'
import {bidWinGenerator} from './bid-win-generator'
import {IEndedState, IWaitingBidState, IWaitingWinState} from './types'

/**
 * Skip n round(s) of game.
 */
export function skip(state: IWaitingBidState | IWaitingWinState, n: number, time: Date): IWaitingBidState | IEndedState {
  if (state.currentRound > state.rounds) {
    // Last round. Return regardless of content of n.
    return toEndedState(state, time)
  }
  if (n <= 0) {
    // Last skip action and not last round
    return toWaitingBidState(state)
  }

  const newState = toWaitingBidState(state)
  newState.bid = bidWinGenerator(Object.keys(state.names))
  newState.scores = mapValues(state.scores, score => [...score, 0])
  newState.currentPlayerOrder = toFront(state.currentPlayerOrder, 1)
  newState.currentRound = state.currentRound + 1

  return skip(newState, n - 1, time)
}
