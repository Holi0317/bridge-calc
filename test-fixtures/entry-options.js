import { deepFreeze } from "./deep-freeze";

export const defaultState = deepFreeze({
  rounds: 13,
  startingRound: 1,
  playerNames: ["John", "Mary", "Henry", "Joe"],
  importOpened: false
});

/**
 * Create mock redux state tree for test related to entry module.
 *
 * @param {object} rest Object to override default state in return tree
 * @returns {object} partial redux tree for testing
 */
export function makeEntryTree(rest) {
  return {
    entry: {
      ...defaultState,
      ...rest
    }
  };
}
