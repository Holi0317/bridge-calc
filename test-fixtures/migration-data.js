export const round1 = {
  'Bridge.currentRound': '1',
  'Bridge.maker': '0',
  'Bridge.players': '[{"bid":null,"win":null,"_buffer":{"lastRound":null,"combo":null,"miss":null,"maxCombo":null,"sum":null},"score":[],"name":"Jonathan"},{"bid":null,"win":null,"_buffer":{"lastRound":null,"combo":null,"miss":null,"maxCombo":null,"sum":null},"score":[],"name":"Helen"},{"bid":null,"win":null,"_buffer":{"lastRound":null,"combo":null,"miss":null,"maxCombo":null,"sum":null},"score":[],"name":"Humphrey"},{"bid":null,"win":null,"_buffer":{"lastRound":null,"combo":null,"miss":null,"maxCombo":null,"sum":null},"score":[],"name":"Margery"}]',
  'Bridge.state': '1',
  'Bridge.totalRounds': '13'
}

export const round2 = {
  'Bridge.currentRound':'2',
  'Bridge.maker':'1',
  'Bridge.players':'[{"bid":null,"win":null,"_buffer":{"lastRound":1,"combo":1,"miss":0,"maxCombo":0,"sum":11},"score":[11],"name":"Jonathan"},{"bid":null,"win":null,"_buffer":{"lastRound":1,"combo":0,"miss":1,"maxCombo":0,"sum":-1},"score":[-1],"name":"Helen"},{"bid":null,"win":null,"_buffer":{"lastRound":1,"combo":0,"miss":1,"maxCombo":0,"sum":-1},"score":[-1],"name":"Humphrey"},{"bid":null,"win":null,"_buffer":{"lastRound":1,"combo":0,"miss":1,"maxCombo":0,"sum":-1},"score":[-1],"name":"Margery"}]',
  'Bridge.state':'1',
  'Bridge.totalRounds':'13'
}
