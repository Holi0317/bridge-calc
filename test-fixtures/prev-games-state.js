export function makePrevGamesTree(prevGames) {
  return {
    prevGames: {
      prevGames
    }
  };
}
