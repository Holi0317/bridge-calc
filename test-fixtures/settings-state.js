import {deepFreeze} from './deep-freeze'

export const defaultState = deepFreeze({
  maker: null,
  names: {}
})
