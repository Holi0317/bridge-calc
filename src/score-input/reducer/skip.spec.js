import {skip} from './skip'
import {waitingBidState, waitingWinState} from '../../../test-fixtures/current-game-states'

// Just being lazy
const time = new Date(1)

test('it should do no-op when n=0', () => {
  const state = {
    ...waitingBidState
  }
  const n = 0
  const expected = {
    ...waitingBidState
  }
  const actual = skip(state, n, time)
  expect(actual).toEqual(expected)
})

test('it should do no-op when n is negative', () => {
  const state = {
    ...waitingBidState
  }
  const n = -5
  const expected = {
    ...waitingBidState
  }
  const actual = skip(state, n, time)
  expect(actual).toEqual(expected)
})

test('it should convert state to waitingBid after skipping', () => {
  const state = {
    ...waitingWinState
  }
  const n = 0
  const expected = {
    ...waitingBidState
  }
  const actual = skip(state, n, time)
  expect(actual).toEqual(expected)
})

test('it should skip rounds', () => {
  const state = {
    ...waitingBidState
  }
  const n = 5
  const actual = skip(state, n, time)
  expect(actual).toMatchSnapshot() // Manually writing expected is too exhausting
})

test('it should end game when n is larger than rounds', () => {
  const state = {
    ...waitingBidState
  }
  const n = 30
  const actual = skip(state, n, time)
  expect(actual).toMatchSnapshot() // Manually writing expected is too exhausting
})

test('it should fill 0 for scores on skipped rounds', () => {
  const state = {
    ...waitingBidState
  }
  const n = 1
  const actual = skip(state, n, time)
  expect(actual).toMatchSnapshot()
})
