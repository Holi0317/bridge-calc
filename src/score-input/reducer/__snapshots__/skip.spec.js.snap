// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should end game when n is larger than rounds 1`] = `
Object {
  "endTime": 1970-01-01T00:00:00.001Z,
  "names": Object {
    "a": "John",
    "b": "Mary",
    "c": "Henry",
    "d": "Joe",
  },
  "rounds": 13,
  "scores": Object {
    "a": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "b": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "c": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "d": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
  "stage": "ended",
  "startTime": 1970-01-01T00:00:00.000Z,
}
`;

exports[`it should fill 0 for scores on skipped rounds 1`] = `
Object {
  "bid": Object {
    "a": 0,
    "b": 0,
    "c": 0,
    "d": 0,
  },
  "currentPlayerOrder": Array [
    "b",
    "c",
    "d",
    "a",
  ],
  "currentRound": 2,
  "names": Object {
    "a": "John",
    "b": "Mary",
    "c": "Henry",
    "d": "Joe",
  },
  "rounds": 13,
  "scores": Object {
    "a": Array [
      0,
    ],
    "b": Array [
      0,
    ],
    "c": Array [
      0,
    ],
    "d": Array [
      0,
    ],
  },
  "stage": "waitingBid",
  "startTime": 1970-01-01T00:00:00.000Z,
}
`;

exports[`it should skip rounds 1`] = `
Object {
  "bid": Object {
    "a": 0,
    "b": 0,
    "c": 0,
    "d": 0,
  },
  "currentPlayerOrder": Array [
    "b",
    "c",
    "d",
    "a",
  ],
  "currentRound": 6,
  "names": Object {
    "a": "John",
    "b": "Mary",
    "c": "Henry",
    "d": "Joe",
  },
  "rounds": 13,
  "scores": Object {
    "a": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "b": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "c": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "d": Array [
      0,
      0,
      0,
      0,
      0,
    ],
  },
  "stage": "waitingBid",
  "startTime": 1970-01-01T00:00:00.000Z,
}
`;
